In [1]:
import os
import sys
import json
from dotenv import load_dotenv

In [2]:
# Load environment variables from .env file
load_dotenv()

True

In [3]:
# Testing YouTube Data API
sys.path.append("../")
from dags import youtube_data_api as yt

### YouTube Data API

In [4]:
DataAPI = yt.LoadDataYT()

In [5]:
# getting channels
channel_params = {
    "part": "snippet,contentDetails,statistics",
    "id": "UC_x5XG1OV2P6uZZ5FSM9Ttw,UCNBmxc6FvKyxtCpUygcdINA",  # comma separated list of ids
    # "maxResults" : 10,
    # "forHandle" : ,

}

channel_info = DataAPI.get_channels(channel_params)

In [6]:
channel_info

{'kind': 'youtube#channelListResponse',
 'etag': '4m4eLfqOlzsSyEXXyiL4XX6g_YE',
 'pageInfo': {'totalResults': 2, 'resultsPerPage': 5},
 'items': [{'kind': 'youtube#channel',
   'etag': 'NaReJdeUVFwL2OQqzJQbrcl4bNk',
   'id': 'UC_x5XG1OV2P6uZZ5FSM9Ttw',
   'snippet': {'title': 'Google for Developers',
    'description': 'Subscribe to join a community of creative developers and learn the latest in Google technology — from AI and cloud, to mobile and web.\n\nExplore more at developers.google.com\n\n',
    'customUrl': '@googledevelopers',
    'publishedAt': '2007-08-23T00:34:43Z',
    'thumbnails': {'default': {'url': 'https://yt3.ggpht.com/2eI1TjX447QZFDe6R32K0V2mjbVMKT5mIfQR-wK5bAsxttS_7qzUDS1ojoSKeSP0NuWd6sl7qQ=s88-c-k-c0x00ffffff-no-rj',
      'width': 88,
      'height': 88},
     'medium': {'url': 'https://yt3.ggpht.com/2eI1TjX447QZFDe6R32K0V2mjbVMKT5mIfQR-wK5bAsxttS_7qzUDS1ojoSKeSP0NuWd6sl7qQ=s240-c-k-c0x00ffffff-no-rj',
      'width': 240,
      'height': 240},
     'high': {'url'

In [7]:
if "items" in channel_info:
    print(len(channel_info["items"]))
    print(1)

2
1


In [8]:
# getting videos
video_params = {
    "part": "snippet,contentDetails,statistics",
    "chart": "mostPopular",
    #"id": "", # comman separated list of ids
    # "myRating": "like",
    # "regionCode": "",
    # "pageToken": "",
    # "maxResults" : 10,
    # when chart is selected we can give videoCategoryId
}

vidoes_list = DataAPI.get_videos(video_params, max_videos=5)

In [9]:
vidoes_list

[{'kind': 'youtube#video',
  'etag': 'KMbTZ4fQRMBkI4KDIhoeu6fp9hY',
  'id': 'SD2UiX1jpd8',
  'snippet': {'publishedAt': '2025-02-22T16:00:06Z',
   'channelId': 'UCooVYzDxdwTtGYAkcPmOgOw',
   'title': 'IS THIS THE BEST BRAWLER EVER?!',
   'description': 'IMPROVED REWARDS FOR F2P, A FULL RANKED REWORK, TWO NEW BRAWLERS, GADGET CHANGES, AND MORE!\nYOUTOOZ PRE-ORDER: https://youtooz.com/collections/brawl-stars\nFREE SPRAY: https://link.brawlstars.com/?action=voucher&code=974a1b0f-b2f7-45c5-978e-bdade83062a1\n\nDownload NOW!! ►► https://supr.cl/2K62CZG\nSubscribe!! ►► https://supr.cl/subscribe\n#brawlstars #brawltalk #update',
   'thumbnails': {'default': {'url': 'https://i.ytimg.com/vi/SD2UiX1jpd8/default.jpg',
     'width': 120,
     'height': 90},
    'medium': {'url': 'https://i.ytimg.com/vi/SD2UiX1jpd8/mqdefault.jpg',
     'width': 320,
     'height': 180},
    'high': {'url': 'https://i.ytimg.com/vi/SD2UiX1jpd8/hqdefault.jpg',
     'width': 480,
     'height': 360},
    'standard': {'

In [10]:
# explore video category list

In [11]:
# getting search results
search_params = {
            'part': 'snippet',
            'q': 'machine learning|Cdeep learning -statistics',  # Searches for "machine learning" OR "deep learning" but NOT "statistics"
            'type': 'channel',
            'maxResults': 2,
            'order': 'videoCount',
            'publishedAfter': '2021-01-01T00:00:00Z',
            #  'pageToken': 
        }

search_results = DataAPI.get_search_results(search_params, max_results=5)

In [12]:
search_results

[{'kind': 'youtube#searchResult',
  'etag': 'Jtxh_TgHRHVsdz_Q-iVKrmOg6QU',
  'id': {'kind': 'youtube#channel', 'channelId': 'UC8F-v_Aq72Z84K7tgYKQyig'},
  'snippet': {'publishedAt': '2024-07-29T18:58:28Z',
   'channelId': 'UC8F-v_Aq72Z84K7tgYKQyig',
   'title': 'Machine learning ',
   'description': 'Plz subscribe my channel.',
   'thumbnails': {'default': {'url': 'https://yt3.ggpht.com/qzLRawvmC9DAvC4D2ierS8eph2jOEtf85UqBDQ0zu17H0T5RhDk3RzN1aHuiCMiyFpI_DpBF=s88-c-k-c0xffffffff-no-rj-mo'},
    'medium': {'url': 'https://yt3.ggpht.com/qzLRawvmC9DAvC4D2ierS8eph2jOEtf85UqBDQ0zu17H0T5RhDk3RzN1aHuiCMiyFpI_DpBF=s240-c-k-c0xffffffff-no-rj-mo'},
    'high': {'url': 'https://yt3.ggpht.com/qzLRawvmC9DAvC4D2ierS8eph2jOEtf85UqBDQ0zu17H0T5RhDk3RzN1aHuiCMiyFpI_DpBF=s800-c-k-c0xffffffff-no-rj-mo'}},
   'channelTitle': 'Machine learning ',
   'liveBroadcastContent': 'none',
   'publishTime': '2024-07-29T18:58:28Z'}},
 {'kind': 'youtube#searchResult',
  'etag': 'J-ybFt9QrSZDAMC_S2lNINEnTSE',
  'id': {'

In [13]:
# getting comments
comment_params = {
    'part': 'snippet',
    'videoId': 'dQw4w9WgXcQ', # comment id ah LoL
    'maxResults': 5,
    'textFormat': 'plainText'
}

comments_list = DataAPI.get_comments(comment_params, max_comments=10)

In [14]:
comments_list

[{'kind': 'youtube#commentThread',
  'etag': 'g9LhHU2ymJL8AodwD-LFuTaIv3I',
  'id': 'UgzarqjaaPC7TbFINNx4AaABAg',
  'snippet': {'channelId': 'UCuAXFkgsw1L7xaCfnd5JJOw',
   'videoId': 'dQw4w9WgXcQ',
   'topLevelComment': {'kind': 'youtube#comment',
    'etag': 'hVvz2ZQfirjX8nezhuc_TUg9p0A',
    'id': 'UgzarqjaaPC7TbFINNx4AaABAg',
    'snippet': {'channelId': 'UCuAXFkgsw1L7xaCfnd5JJOw',
     'videoId': 'dQw4w9WgXcQ',
     'textDisplay': '1 BILLION views for Never Gonna Give You Up!\xa0 Amazing, crazy, wonderful! Rick ♥️',
     'textOriginal': '1 BILLION views for Never Gonna Give You Up!\xa0 Amazing, crazy, wonderful! Rick ♥️',
     'authorDisplayName': '@RickAstleyYT',
     'authorProfileImageUrl': 'https://yt3.ggpht.com/K2ecE5j90a_DFzugHo0bW98vFlIQ1JJgs9mbcav7RGy1t7adJRnd2jaIv-oc6XzTRvDdWlFCAfc=s48-c-k-c0x00ffffff-no-rj',
     'authorChannelUrl': 'http://www.youtube.com/@RickAstleyYT',
     'authorChannelId': {'value': 'UCuAXFkgsw1L7xaCfnd5JJOw'},
     'canRate': True,
     'viewerRati

In [30]:
# getting comments
comment_params = {
    'part': 'id,snippet,replies',
    'allThreadsRelatedToChannelId': "UC_x5XG1OV2P6uZZ5FSM9Ttw", # comment id ah LoL
    #'id': 'UgyS5DZZTo4_ebj5Ru54AaABAg',
    'maxResults': 5,
    'textFormat': 'plainText'
}

comments_list = DataAPI.get_comments(comment_params, max_comments=10)

In [31]:
comments_list

[{'kind': 'youtube#commentThread',
  'etag': 'iimp086bqV8PTmHtw9FhMEbs2rg',
  'id': 'UgyS5DZZTo4_ebj5Ru54AaABAg',
  'snippet': {'channelId': 'UC_x5XG1OV2P6uZZ5FSM9Ttw',
   'videoId': 'LQkqBYOREgE',
   'topLevelComment': {'kind': 'youtube#comment',
    'etag': 'TTifuFzeq5_rroJpHIPlU3Re6Kg',
    'id': 'UgyS5DZZTo4_ebj5Ru54AaABAg',
    'snippet': {'channelId': 'UC_x5XG1OV2P6uZZ5FSM9Ttw',
     'videoId': 'LQkqBYOREgE',
     'textDisplay': "Wagtail is really really great.  Anyone outside the Django community that doesn't know about it is missing out",
     'textOriginal': "Wagtail is really really great.  Anyone outside the Django community that doesn't know about it is missing out",
     'authorDisplayName': '@ianokay',
     'authorProfileImageUrl': 'https://yt3.ggpht.com/ytc/AIdro_ngIMlirWw1eEdyXnSD2HjV1JCQJFgOOQnQT0dzOp6dzalDb4MvxXxuXMw-KT5R6ckjgQ=s48-c-k-c0x00ffffff-no-rj',
     'authorChannelUrl': 'http://www.youtube.com/@ianokay',
     'authorChannelId': {'value': 'UCpsg2LcpQaIDZLNOoX

In [17]:
# getting captions
caption_params = {
    'part': 'snippet',
    'videoId': 'ERCMXc8x7mc',
}

captions_data = DataAPI.get_captions(caption_params)

In [18]:
captions_data

# captions are useless

[{'kind': 'youtube#caption',
  'etag': 'Jw88JJGDxJRI1rFGDKMREv2PJ44',
  'id': 'AUieDabhtdjibZLKzI7FsBgTpcopA3VCEzYQMMU5SJb7kiDfojg',
  'snippet': {'videoId': 'ERCMXc8x7mc',
   'lastUpdated': '2024-09-21T20:22:13.728894Z',
   'trackKind': 'asr',
   'language': 'hi',
   'name': '',
   'audioTrackType': 'unknown',
   'isCC': False,
   'isLarge': False,
   'isEasyReader': False,
   'isDraft': False,
   'isAutoSynced': False,
   'status': 'serving'}}]

### YouTube Reporting and Analytics APIs

In [19]:
AnRAPI = yt.LoadAR()

In [20]:
# retrieve analytics report
# anal_params = {
#     'ids': 'channel==UCNBmxc6FvKyxtCpUygcdINA',
#     'startDate': '2024-01-01',
#     'endDate': '2024-12-31',
#     'metrics': 'views,likes,dislikes,comments',
#     'dimensions': 'day',  # Aggregate data by day
#     #'filters': 'country==US',  # Filter by country, video==ID;
#     'sort': '-views'  # Sort by descending views
#     # 'includeHistoricalChannelData'
#     # 'maxResults'
# }

# get_anal_report = AnRAPI.retrieve_analytics_report(anal_params)

# not possible for other youtube channels data


In [21]:
# create a reporting job
report_type_id = 'contentOwnerMonetizationReport'
name = 'Monetization Report for 2024'
onBehalfOfContentOwner = "ID content owner - default mine"

# creatTime, expireTime, id, name, systemManaged

# need to give a job body


In [22]:
# list reporting jobs


In [23]:
# downloading reports
job_id = 'job12345'
report_id = 'report67890'
download_path = '/path/to/save/report.csv'

What a bummer, I can't actually use YouTube reporting and analytics tools...

Oh god!!!

But it will be damn useful...

[{'kind': 'youtube#searchResult',
  'etag': 'lEVPJB-1IwuyV8gFOWKCQvhnntg',
  'id': {'kind': 'youtube#video', 'videoId': 'ERCMXc8x7mc'},
  'snippet': {'publishedAt': '2024-02-12T14:24:29Z',
   'channelId': 'UCBwmMxybNva6P_5VmxjzwqA',
   'title': 'Python Tutorial for Beginners - Full Course (with Notes &amp; Practice Questions)',
   'description': 'You can start your Tech Placement/Internship journey with New Sigma 4.0 : https://www.apnacollege.in/course/sigma-4 Early bird ...',
   'thumbnails': {'default': {'url': 'https://i.ytimg.com/vi/ERCMXc8x7mc/default.jpg',
     'width': 120,
     'height': 90},
    'medium': {'url': 'https://i.ytimg.com/vi/ERCMXc8x7mc/mqdefault.jpg',
     'width': 320,
     'height': 180},
    'high': {'url': 'https://i.ytimg.com/vi/ERCMXc8x7mc/hqdefault.jpg',
     'width': 480,
     'height': 360}},
   'channelTitle': 'Apna College',
   'liveBroadcastContent': 'none',
   'publishTime': '2024-02-12T14:24:29Z'}},
 {'kind': 'youtube#searchResult',
  'etag': 'UM1kKaLI96NY4tLErrEhnrAtO6Q',
  'id': {'kind': 'youtube#video', 'videoId': 'UrsmFxEIp5k'},
  'snippet': {'publishedAt': '2024-06-16T05:36:24Z',
   'channelId': 'UCeVMnSShP_Iviwkknt83cww',
   'title': 'Python Tutorial For Beginners in Hindi | Complete Python Course 🔥',
   'description': 'Source Code: https://github.com/CodeWithHarry/The-Ultimate-Python-Course Python PDF Handbook (authored by me): ...',
   'thumbnails': {'default': {'url': 'https://i.ytimg.com/vi/UrsmFxEIp5k/default.jpg',
     'width': 120,
     'height': 90},
    'medium': {'url': 'https://i.ytimg.com/vi/UrsmFxEIp5k/mqdefault.jpg',
     'width': 320,
     'height': 180},
    'high': {'url': 'https://i.ytimg.com/vi/UrsmFxEIp5k/hqdefault.jpg',
     'width': 480,
     'height': 360}},
   'channelTitle': 'CodeWithHarry',
   'liveBroadcastContent': 'none',
   'publishTime': '2024-06-16T05:36:24Z'}},
 {'kind': 'youtube#searchResult',
  'etag': 'bExSUiYXlyWxJ_rttjwTZMTGtbQ',
  'id': {'kind': 'youtube#video', 'videoId': 't2_Q2BRzeEE'},
  'snippet': {'publishedAt': '2024-01-26T15:31:21Z',
   'channelId': 'UC1XBh-m27kkgwLAwu_SRJBg',
   'title': 'Python Full Course❤️ | Variables &amp; Data Types | Lecture 1',
   'description': 'This lecture was made with a lot of love❤️ Notes ...',
   'thumbnails': {'default': {'url': 'https://i.ytimg.com/vi/t2_Q2BRzeEE/default.jpg',
     'width': 120,
     'height': 90},
    'medium': {'url': 'https://i.ytimg.com/vi/t2_Q2BRzeEE/mqdefault.jpg',
     'width': 320,
     'height': 180},
    'high': {'url': 'https://i.ytimg.com/vi/t2_Q2BRzeEE/hqdefault.jpg',
     'width': 480,
     'height': 360}},
   'channelTitle': 'Shradha Khapra',
   'liveBroadcastContent': 'none',
   'publishTime': '2024-01-26T15:31:21Z'}},
 {'kind': 'youtube#searchResult',
  'etag': 'sphQz-viso8DCr7pp_u5cgbY8OU',
  'id': {'kind': 'youtube#video', 'videoId': 'wjZofJX0v4M'},
  'snippet': {'publishedAt': '2024-04-01T19:13:57Z',
   'channelId': 'UCYO_jab_esuFRV4b17AJtAw',
   'title': 'Transformers (how LLMs work) explained visually | DL5',
   'description': 'Breaking down how Large Language Models work Instead of sponsored ad reads, these lessons are funded directly by viewers: ...',
   'thumbnails': {'default': {'url': 'https://i.ytimg.com/vi/wjZofJX0v4M/default.jpg',
     'width': 120,
     'height': 90},
    'medium': {'url': 'https://i.ytimg.com/vi/wjZofJX0v4M/mqdefault.jpg',
     'width': 320,
     'height': 180},
    'high': {'url': 'https://i.ytimg.com/vi/wjZofJX0v4M/hqdefault.jpg',
     'width': 480,
     'height': 360}},
   'channelTitle': '3Blue1Brown',
   'liveBroadcastContent': 'none',
   'publishTime': '2024-04-01T19:13:57Z'}},
 {'kind': 'youtube#searchResult',
  'etag': '_EbFmS9_EfDNpSeZSqqNDBhJ-Yo',
  'id': {'kind': 'youtube#video', 'videoId': 'VaSjiJMrq24'},
  'snippet': {'publishedAt': '2024-01-07T13:30:10Z',
   'channelId': 'UC0T6MVd3wQDB5ICAe45OxaQ',
   'title': 'Data Analytics FULL Course for Beginners to Pro in 29 HOURS - 2025 Edition',
   'description': 'Data Analytics FULL Course for Beginners to Pro in 29 HOURS - 2025 Edition To learn Data Analytics Course online with ...',
   'thumbnails': {'default': {'url': 'https://i.ytimg.com/vi/VaSjiJMrq24/default.jpg',
     'width': 120,
     'height': 90},
    'medium': {'url': 'https://i.ytimg.com/vi/VaSjiJMrq24/mqdefault.jpg',
     'width': 320,
     'height': 180},
    'high': {'url': 'https://i.ytimg.com/vi/VaSjiJMrq24/hqdefault.jpg',
     'width': 480,
     'height': 360}},
   'channelTitle': 'WsCube Tech',
   'liveBroadcastContent': 'none',
   'publishTime': '2024-01-07T13:30:10Z'}}]
